File: rf2awsiot.py
#!/usr/bin/env python
"""
rf2awsiot.py v1 Serial to AWS interface 
---------------------------------------------------------------------------------
                                                                                  
 J. Evans June 2018
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
 WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                                       
                                                                                  
 Revision History                                                                  
 V1.00 - Release
"""

import serial
import globals
import time
import sys
import thread
from AWSIoTPythonSDK.MQTTLib import AWSIoTMQTTShadowClient
import logging
import json
import argparse
from time import sleep
global RFList
global LocationList
global numrf
DEBUG = True
Fahrenheit = False

# Custom Shadow callback
def customShadowCallback_Update(payload, responseStatus, token):
    # payload is a JSON string ready to be parsed using json.loads(...)
    # in both Py2.x and Py3.x
    if responseStatus == "timeout":
        print("Update request " + token + " time out!")
    if responseStatus == "accepted":
        print("~~~~~~~~~~~~~~~~~~~~~~~")
        print("Update request with token: " + token + " accepted!")
        print("property: " + str(payload))
        print("~~~~~~~~~~~~~~~~~~~~~~~\n\n")
    if responseStatus == "rejected":
        print("Update request " + token + " rejected!")

def customShadowCallback_Delete(payload, responseStatus, token):
    if responseStatus == "timeout":
        print("Delete request " + token + " time out!")
    if responseStatus == "accepted":
        print("~~~~~~~~~~~~~~~~~~~~~~~")
        print("Delete request with token: " + token + " accepted!")
        print("~~~~~~~~~~~~~~~~~~~~~~~\n\n")
    if responseStatus == "rejected":
        print("Delete request " + token + " rejected!")

def dprint(message):
	if (DEBUG):
		print message

def ProcessMessageThread(value, value2, DevId, type):
	try:
			thread.start_new_thread(ProcessMessage, (value, value2, DevId, type, ) )
	except:
			print "Error: unable to start thread"

def aws_send(device_id, value, property):
	JSONPayload = '{"state":{"reported":{"'+str(device_id)+'": {"'+str(property)+'": "'+ str(value) + '"}}}}'
	dprint(JSONPayload)
	deviceShadowHandler.shadowUpdate(JSONPayload, customShadowCallback_Update, 5)
			
def ProcessMessage(value, value2, DevId, type, property):
# Notify the host that there is new data from a sensor (e.g. door open)
	try:
		dprint("Processing data : DevId="+str(DevId)+",Type="+str(type)+",Value1="+str(value)+",Value2="+str(value2))

		DevId="Device_"+DevId;
		#Send switch sensor value to host
		if type==1:
				value=value[1:]
				if value=='OF' or value=='OFF':
						aws_send(DevId, "Open", property)
				if value=='ON':
						aws_send(DevId, "Closed", property)

		#Send battery level to host
		if type==2:
				aws_send(DevId, value, property)

		#Send temperature to host
		if type==3:
				if Fahrenheit:
						value = value*1.8+32
						value = round(value,2)
				
				aws_send(str(DevId), str(value), property)

		#Send humidity to host
		if type==4:
				if Fahrenheit:
						value = value*1.8+32
						value = round(value,2)
				
				aws_send(DevId, str(value), "TMP")
				aws_send(DevId, str(value2), "HUM")
				
	except Exception as e: dprint(e)
	return(0)

def main():
        currvalue=''
        tempvalue=-999;
        
        # loop until the serial buffer is empty

        start_time = time.time()
        
        #try:
        while True:

				# declare to variables, holding the com port we wish to talk to and the speed
				port = '/dev/ttyAMA0'
				baud = 9600

				# open a serial connection using the variables above
				ser = serial.Serial(port=port, baudrate=baud)

				# wait for a moment before doing anything else
				sleep(0.2)        
				while ser.inWaiting():
						# read a single character
						char = ser.read()
						# check we have the start of a LLAP message
						if char == 'a':
								sleep(0.01)
								start_time = time.time()
								
								# start building the full llap message by adding the 'a' we have
								llapMsg = 'a'

								# read in the next 11 characters form the serial buffer
								# into the llap message
								llapMsg += ser.read(11)

								# now we split the llap message apart into devID and data
								devID = llapMsg[1:3]
								data = llapMsg[3:]
								
								dprint(time.strftime("%c")+ " " + llapMsg)
																
								if data.startswith('BUTTON'):
										sensordata=data[5:].strip('-')
										if currvalue<>sensordata or currvalue=='':
												currvalue=sensordata
												ProcessMessage(currvalue, 0, devID,1, "BUTTON")

								if data.startswith('BTN'):
										sensordata=data[2:].strip('-')
										if currvalue<>sensordata or currvalue=='':
												currvalue=sensordata
												ProcessMessage(currvalue, 0, devID,1, "BUTTON")

								if data.startswith('TMPA'):
										sensordata=str(data[4:].rstrip("-"))								
										currvalue=sensordata
										ProcessMessage(currvalue, 0, devID,3, "TMPA")
								
								if data.startswith('ANAA'):
										sensordata=str(data[4:].rstrip("-"))								
										currvalue=sensordata
										ProcessMessage(currvalue, 0, devID,3, "ANAA")
								
								if data.startswith('ANAB'):
										sensordata=str(data[4:].rstrip("-"))								
										currvalue=sensordata
										ProcessMessage(currvalue, 0, devID,3, "ANAB")
								
								if data.startswith('TMPC'):
										sensordata=str(data[4:].rstrip("-"))								
										currvalue=sensordata
										ProcessMessage(currvalue, 0, devID,3, "TMPC")
								
								if data.startswith('TMPB'): #Temperature followed by humidity
										sensordata=str(data[4:].rstrip("-"))								
										tempbdata=sensordata
																				
								if data.startswith('HUM'):
										sensordata=str(data[3:].rstrip("-"))								
										currvalue=sensordata
										if tempbdata<>"" and sensordata<>"":
											ProcessMessage(tempbdata, sensordata, devID,4, "")
											tempbdata=''
												
								# check if battery level is being sent axxBATTn.nn-
								if data.startswith('BATT'):
										sensordata=data[4:].strip('-')
										currvalue=sensordata 
										ProcessMessage(currvalue, 0, devID,2, "BATT")
				elapsed_time = time.time() - start_time
				if (elapsed_time > 2):
						currvalue=""
						sensordata=""
						tempbdata=""
				sleep(0.2)
				#except:
				#        print "Error: unable to start thread"
           
if __name__ == "__main__":
        # Read in command-line parameters
		parser = argparse.ArgumentParser()
		parser.add_argument("-e", "--endpoint", action="store", required=True, dest="host", help="Your AWS IoT custom endpoint")
		parser.add_argument("-r", "--rootCA", action="store", required=True, dest="rootCAPath", help="Root CA file path")
		parser.add_argument("-c", "--cert", action="store", dest="certificatePath", help="Certificate file path")
		parser.add_argument("-k", "--key", action="store", dest="privateKeyPath", help="Private key file path")
		parser.add_argument("-w", "--websocket", action="store_true", dest="useWebsocket", default=False,help="Use MQTT over WebSocket")
		parser.add_argument("-n", "--thingName", action="store", dest="thingName", default="Bot", help="Targeted thing name")
		parser.add_argument("-id", "--clientId", action="store", dest="clientId", default="basicShadowUpdater", help="Targeted client id")

		args = parser.parse_args()
		host = args.host
		rootCAPath = args.rootCAPath
		certificatePath = args.certificatePath
		privateKeyPath = args.privateKeyPath
		useWebsocket = args.useWebsocket
		thingName = args.thingName
		clientId = args.clientId

		if args.useWebsocket and args.certificatePath and args.privateKeyPath:
			parser.error("X.509 cert authentication and WebSocket are mutual exclusive. Please pick one.")
			exit(2)

		if not args.useWebsocket and (not args.certificatePath or not args.privateKeyPath):
			parser.error("Missing credentials for authentication.")
			exit(2)

		# Configure logging
		logger = logging.getLogger("AWSIoTPythonSDK.core")
		logger.setLevel(logging.DEBUG)
		streamHandler = logging.StreamHandler()
		formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
		streamHandler.setFormatter(formatter)
		logger.addHandler(streamHandler)

		# Init AWSIoTMQTTShadowClient
		myAWSIoTMQTTShadowClient = None
		if useWebsocket:
			myAWSIoTMQTTShadowClient = AWSIoTMQTTShadowClient(clientId, useWebsocket=True)
			myAWSIoTMQTTShadowClient.configureEndpoint(host, 443)
			myAWSIoTMQTTShadowClient.configureCredentials(rootCAPath)
		else:
			myAWSIoTMQTTShadowClient = AWSIoTMQTTShadowClient(clientId)
			myAWSIoTMQTTShadowClient.configureEndpoint(host, 8883)
			myAWSIoTMQTTShadowClient.configureCredentials(rootCAPath, privateKeyPath, certificatePath)

		# AWSIoTMQTTShadowClient configuration
		myAWSIoTMQTTShadowClient.configureAutoReconnectBackoffTime(1, 32, 20)
		myAWSIoTMQTTShadowClient.configureConnectDisconnectTimeout(10)  # 10 sec
		myAWSIoTMQTTShadowClient.configureMQTTOperationTimeout(5)  # 5 sec

		# Connect to AWS IoT
		myAWSIoTMQTTShadowClient.connect()

		# Create a deviceShadow with persistent subscription
		deviceShadowHandler = myAWSIoTMQTTShadowClient.createShadowHandlerWithName(thingName, True)

		# Delete shadow JSON doc
		deviceShadowHandler.shadowDelete(customShadowCallback_Delete, 5)
		main()



   
   

File: rf2awsiot.js
/*rf2awsiot.js Interface between RF Module serial interface and Amazon AWS IOT Core
---------------------------------------------------------------------------------                                                                               
 J. Evans May 2018
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
 WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                                       
 
 Revision History                                                                  
 V1.00 - Release
 
 -----------------------------------------------------------------------------------
*/
 
'use strict';
var awsIot = require('aws-iot-device-sdk');

//Set Fahrenheit=0 display in centigrade
const Fahrenheit=0;

//Set the AWS Thing Prefix
const thingPrefix = "Device_";
const thingName = "RaspberryPi";

//Set the AWS End Point e.g. a1jj45tr9rhhbq.iot.us-west-2.amazonaws.com
const endPoint = 'a1j34fr34frhhbq.iot.us-west-2.amazonaws.com';

// Replace the values below with a unique client identifier and custom host endpoint provided 
// in AWS IoT cloud NOTE: client identifiers must be unique within your AWS account; if a client 
// attempts to connect with a client identifier which is already in use, the existing connection 
// will be terminated.

var thingShadows = awsIot.thingShadow({
   keyPath: '/home/pi/aws-keys/private.pem.key',
  certPath: '/home/pi/aws-keys/certificate.pem.crt',
    caPath: '/home/pi/aws-keys/root-CA.crt',
  clientId: thingName,
      host: endPoint
});

//Library used for serial port access
var SerialPort = require('serialport');
var port = new SerialPort('/dev/ttyAMA0');

var inStr="";

var data = {
	deviceID: "",
    command: "",
    value: ""
};

//
// Client token value returned from thingShadows.update() operation
//
var clientTokenUpdate;

thingShadows.on('connect', function() {
	console.log('AWS Connected');	
});

thingShadows.on('error', function(err) {
  console.log('Error: ', err.message);
});

thingShadows.on('status', 
    function(thingName, stat, clientToken, stateObject) {
       console.log('received '+stat+' on '+thingName+': '+
                   JSON.stringify(stateObject));
//
// These events report the status of update(), get(), and delete() 
// calls.  The clientToken value associated with the event will have
// the same value which was returned in an earlier call to get(),
// update(), or delete().  Use status events to keep track of the
// status of shadow operations.
//
    });

thingShadows.on('delta', 
    function(thingName, stateObject) {
       console.log('received delta on '+thingName+': '+
                   JSON.stringify(stateObject));
    });

thingShadows.on('timeout',
    function(thingName, clientToken) {
       console.log('received timeout on '+thingName+
                   ' with token: '+ clientToken);
//
// In the event that a shadow operation times out, you'll receive
// one of these events.  The clientToken value associated with the
// event will have the same value which was returned in an earlier
// call to get(), update(), or delete().
//
    });
	
// Read data that is available but keep the stream from entering "flowing mode"
port.on('readable', function () {
  var n;
  var deviceID;
  var payload;
  var llapMsg;
  
  inStr+=port.read().toString('utf8');
  n = inStr.search("a"); //start charachter for llap message
  if (n>0) inStr = inStr.substring(n, inStr.length); //chop off data preceding start charachter
  if (inStr.length>=12){ //we have an llap message!
    while (inStr!=""){
		data.command="";
		llapMsg=inStr.substring(1,12);
		console.log(llapMsg);
		data.deviceID=llapMsg.substring(0,2);
		deviceID=thingPrefix+data.deviceID;
		if (llapMsg.substring(2,6)=="TMPA") {
			data.value=llapMsg.substring(6,13);
			data.command="TMPA";
		}
		if (llapMsg.substring(2,6)=="TMPB") {
			data.value=llapMsg.substring(6,13);
			data.command="TMPB";
		}
		if (llapMsg.substring(2,5)=="HUM") {
			data.value=llapMsg.substring(5,13);
			data.command="HUM";
		}
		if (llapMsg.substring(2,6)=="TMPC") {
			data.value=llapMsg.substring(6,13);
			data.command="TMPC";
		}
		if (llapMsg.substring(2,8)=="BUTTON") {
			data.value=llapMsg.substring(8,13);
			data.command="BUTTON";
		}
		if (llapMsg.substring(2,10)=="SLEEPING") {
			data.value="ASLEEP";
			data.command="STATE";
		}
	    if (llapMsg.substring(2,7)=="AWAKE") {
			data.value="AWAKE";
			data.command="STATE";
		}
		if (llapMsg.substring(2,6)=="BATT") {
			data.value=llapMsg.substring(6,13);
			data.command="BATT";
		}
		if (llapMsg.substring(2,9)=="STARTED"){
			data.value="";
			data.command="STARTED";
		} 
		if (data.command!="" && data.command!="STATE"){ 
			data.value.replace('-',' ');
			data.value.trim();
			if (Fahrenheit){
				if (data.command=="TMP"){
					data.value=data.value*1.8+32;
					data.value=data.value.toFixed(2);
				}
			}
				
			var awsData =  "{\"state\":{\"reported\":{\""+deviceID+"\": {\""+data.command+"\":\""+data.value+"\"}}}}";
			var awsJSON = JSON.parse(awsData);
			console.log(awsJSON);
			
			//
			// After connecting to the AWS IoT platform, register interest in the
			// Thing Shadow named 'RaspberryPi', for example.
			//
			
			// Once registration is complete, update the Thing Shadow named
			// 'RaspberryPi', for example, with the latest device state and save the clientToken
			// so that we can correlate it with status or timeout events.
			//
			// Thing shadow state
			//
			
			if (data.command=="TMP"){ //Set to only send temperature reading to AWS. Remove this and all will get sent
				console.log("Registering Thing:"+thingName);
			    thingShadows.register(thingName);
				clientTokenUpdate = thingShadows.update(thingName, awsJSON);
				//
				// The update method returns a clientToken; if non-null, this value will
				// be sent in a 'status' event when the operation completes, allowing you
				// to know whether or not the update was successful.  If the update method
				// returns null, it's because another operation is currently in progress and
				// you'll need to wait until it completes (or times out) before updating the 
				// shadow.
				//
				if (clientTokenUpdate === null)
					{
					console.log('update shadow failed, operation still in progress');
					}
				}				
			}
		if (inStr.length>12) 
			inStr=inStr.substring(12,inStr.length);
		else
			inStr="";
	  }  
  }
});


port.on('error', function(err) {
  console.log('Error: ', err.message);
});
